Adding chess timer

alexander-novo opened the pull request:
Adding in a chess timer with a "buffer time" which is used before a player's normal time. This buffer time is refreshed every time the timer is invoked, similar to a chess clock. This way players can lower the normal player game time to prevent players from stalling out the game for too long without fear of a long game with many game actions running out their normal clocks.

Some thoughts on WIP stuff that I'd like people's opinions on:

- Current options for Buffer Time are just 5, 10, 15 seconds. Do we need anything longer than this? Or more granular options?
- The default match settings currently are the same game time and no buffer time. 
- I don't think I'm going to add a UI element for the buffer timer in game. It seems like too much work for what it's worth.

![image](https://github.com/magefree/mage/assets/11657716/876d3422-fbce-4e54-8dbc-fbf93b8027da)

![image](https://github.com/magefree/mage/assets/11657716/cac21edf-cd67-44d4-bb43-913036e5d5ef)

Closes #10597

JayDi85 commented:
I don't think it's need a GUI at all. Must be default for any games.

alexander-novo commented:
You think? I worry that in that case it will end up inaccessible and unchanged when it needs to be like a lot of those features for the BETA server.

xenohedron commented:
I like this. Great functionality to have as an option, and I agree that it must be easily configurable on table/tournament creation as folks will be used to old fixed timer settings and this will be an adjustment. For now I think 5/10/15 seconds is fine, can always be adjusted later (and/or made default) based on feedback from actual usage.

alexander-novo commented:
Okay I thought of a good way to do the UI that would be simple to do, so I went and added one. I added the buffer time to the player's timer (so every time their action starts, their timer will increase up to the amount of buffer time the game has been configured for) and if the player is currently using their buffer time, then the timer text is green:

https://github.com/magefree/mage/assets/11657716/b70fd4d9-c927-4de6-94a0-425489e20298

xenohedron commented on Mage.Client/src/main/java/mage/client/dialog/NewTableDialog.form:
```diff
@@ -364,6 +364,14 @@
+        <Property name="toolTipText" type="java.lang.String" value="The extra time a player gets whenever the timer starts before their normal time limit starts going down."/>
+      </Properties>
+    </Component>
+    <Component class="javax.swing.JComboBox" name="cbBufferTime">
```
The NewTournamentDialog sets a tooltip here, but this NewTableDialog doesn't. Is this intentional or should one of them be changed?

alexander-novo replied:
This is consistent with the prior implementation for the time limit option. I honestly don't think a combo box needs a tooltip - it makes more sense to hover the label.

xenohedron commented on Mage.Client/src/main/java/mage/client/dialog/NewTournamentDialog.form:
```diff
@@ -373,6 +373,20 @@
+    </Component>
+    <Component class="javax.swing.JComboBox" name="cbBufferTime">
+      <Properties>
+        <Property name="toolTipText" type="java.lang.String" value="The time a player gets whenever their timer starts before their match time starts going down. "/>
```
See other review comment re: tooltip discrepancy

xenohedron reviewed (approved):
Code looks good to me, just one small question.

Anything else you wanted to do with this before merge?

alexander-novo commented:
No, I think I am content with the way it works right now.

artemiswkearney commented:
@alexander-novo
In the future, I suggest you heed this warning:
```java
    /**
     * This method is called from within the constructor to initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is always
     * regenerated by the Form Editor.
     */
```
From what I can tell, you made your UI changes by directly editing the `.form` files and generated code sections. While this appears to work (and does a good enough impression that xenohedron was convinced), the new elements were never given any positioning info in the form, meaning they **disappear from the dialogs entirely** the next time those sections are regenerated (which happens any time anyone makes UI changes the correct way).

We're still figuring out how to handle this, given that the feature has already been merged and built into a release. What's probably going to have to happen is that I (as one of the few people willing to actually install Netbeans 8) will have to redo your UI work the right way, get that merged, and only then resume my work on #10551.

If you need to make any GUI changes in the future, please either use the Netbeans form editor or ask for help from someone who already has it set up. I'm reachable in the xmage-dev channel on the XDHS Discord server, and there are probably others on the main XMage server as well.
